File: deep_clean.R

# Script to clean raw DOGE data to the University of Maryland

# library
library(tidyverse)

# Reading in raw DOGE data
doge_data <- read_csv("data/doge_data.csv")

# Cleaning the data
clean_doge_data <- doge_data |>
  mutate(
    savings = str_extract(details, "\\$?[0-9,\\.]+(?=\\s+[Ss]avings)"),
    contract_total = str_extract(details, "\\$?[0-9,\\.]+(?=\\s+[Tt]otal)"),
    contract_description = str_extract(details, "(?i)Total Contract\\s+.*") |> 
      str_remove("(?i)Total Contract\\s+"),
    grant_description = str_extract(details, "(?i)Total Grant\\s+.*") |> 
      str_remove("(?i)Total Grant\\s+"),
    providing_agency = str_extract(details, "(?i)Agency:\\s*[^$]+") |>
      str_remove("(?i)Agency:\\s*"),
    vendor = str_extract(details, "^(.*?)\\s*[Aa]gency:") |>
      str_remove("\\s*[Aa]gency:")
  )

clean_doge_data <- clean_doge_data |>
  mutate(
    vendor = str_to_lower(vendor),
    full_details = str_to_lower(details)
  )

# writing to the csv
write_csv(umd_doge_cuts, "data/clean_cuts.csv")

File: deep_clean.py
import pandas as pd


def extract_details(df, col="details"):
	s = df[col].astype(str)
	df = df.copy()
	df["savings"] = s.str.extract(r"(\$?[0-9,\.]+)(?=\s+[Ss]avings)", expand=False)
	df["contract_total"] = s.str.extract(r"(\$?[0-9,\.]+)(?=\s+[Tt]otal)", expand=False)
	df["contract_description"] = s.str.extract(r"(?i)Total Contract\s+(.*)", expand=False)
	df["grant_description"] = s.str.extract(r"(?i)Total Grant\s+(.*)", expand=False)
	df["providing_agency"] = s.str.extract(r"(?i)Agency:\s*([^$]+)", expand=False)
	df["vendor"] = s.str.extract(r"^(.*?)\s*[Aa]gency:", expand=False)
	df["vendor"] = df["vendor"].str.lower()
	df["full_details"] = s.str.lower()
	# strip whitespace from some extracted fields
	for c in ["contract_description", "grant_description", "providing_agency", "vendor"]:
		if c in df.columns:
			df[c] = df[c].astype(str).str.strip().replace({'nan': None})
	return df


def main():

	df = pd.read_csv("data/doge_data.csv")
	df_clean = extract_details(df, col="details")
	# drop the original 'details' column per request
	df_clean = df_clean.drop(columns=["details"], errors="ignore")
	df_clean.to_csv("data/clean_cuts.csv", index=False)


if __name__ == "__main__":
	main()
